### PR TITLE
qa(metrics): testing

### PR DESCRIPTION
**Pull request by aarnphm**
Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

metrics should all be tested now.

### CODE REVIEW

File: bentoml/_internal/external_typing/prometheus.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
+from __future__ import annotations
```
Comment (parano) [inline bentoml/_internal/external_typing/prometheus.py:1]: do we really need this now we have turned on `analysis.useLibraryCodeForTypes = true`?

File: bentoml/_internal/utils/analytics/usage_stats.py [hunk 2]
```diff
@@ -36,9 +38,11 @@
 
 BENTOML_DO_NOT_TRACK = "BENTOML_DO_NOT_TRACK"
 USAGE_TRACKING_URL = "https://t.bentoml.com"
-SERVE_USAGE_TRACKING_INTERVAL_SECONDS = int(12 * 60 * 60)  # every 12 hours
 USAGE_REQUEST_TIMEOUT_SECONDS = 1
 
+# Send a serve event every 12 hours
+_serve_usage_tracking_interval_seconds = int(12 * 60 * 60)
```
Comment (parano) [inline bentoml/_internal/utils/analytics/usage_stats.py:44]: why renaming? use all cap for constants

Comment (aarnphm) [reply bentoml/_internal/utils/analytics/usage_stats.py:44]: oops

Comment (aarnphm) [reply bentoml/_internal/utils/analytics/usage_stats.py:44]: I think my refactoring library have some issues

File: bentoml/_internal/external_typing/prometheus.py (hunk 1 above)
Comment (aarnphm) [reply bentoml/_internal/external_typing/prometheus.py:1]: it doesn't work with `analysis.useLibraryCodeForTypes` when the library doesn't even have type hint at the first place :)

File: bentoml/_internal/utils/circus.py [hunk 3]
```diff
@@ -89,18 +99,23 @@ def __init__(self, *args: t.Any, **config: t.Any):
 
         # circus/plugins/__init__.py:282 -> converts all given configs to dict[str, str]
         self.reload_delay: float = float(self.config.get("reload_delay", 1))
-        self.file_watcher = PyFileChangeWatcher([working_dir])
+        self.reloader = BentoReload(working_dir=working_dir)
```
Comment (bojiang) [inline bentoml/_internal/utils/circus.py:102]: The name `xxx_watcher` is still better in my opinion

File: bentoml/_internal/utils/circus.py [hunk 4]
```diff
@@ -89,18 +99,23 @@ def __init__(self, *args: t.Any, **config: t.Any):
...
-            self.file_watcher.reset()
+        for changes in self.reloader:
+            if changes:
+                logger.warning(
+                    "%s detected changes in %s. Reloading...",
+                    self.__class__.__name__,
+                    ", ".join(map(display_path, changes)),
+                )
+            break
+        self.call("restart", name="*")  # type: ignore
```
Comment (bojiang) [inline bentoml/_internal/utils/circus.py:113]: why do we call restart for every look_after?

File: bentoml/_internal/utils/uri.py [hunk 5]
```diff
@@ -7,7 +10,7 @@
 import psutil
 
 
-def path_to_uri(path: str) -> str:
+def path_to_uri(path: str) -> str | t.NoReturn:
```
Comment (bojiang) [inline bentoml/_internal/utils/uri.py:13]:
```suggestion
def path_to_uri(path: str) -> str:
```

File: bentoml/_internal/server/metrics/prometheus.py [hunk 6]
```diff
@@ -83,41 +103,59 @@ def mark_process_dead(self) -> None:
...
+    def generate_latest(self) -> bytes:
         if self.multiproc:
             registry = self.prometheus_client.CollectorRegistry()
             self.prometheus_client.multiprocess.MultiProcessCollector(registry)
-            return self.prometheus_client.generate_latest(registry)
+            return t.cast(bytes, self.prometheus_client.generate_latest(registry))
         else:
             return self.prometheus_client.generate_latest()
 
     def get_metrics_report(
```
Comment (bojiang) [inline bentoml/_internal/server/metrics/prometheus.py:120]: nits: shall we maintain the analytics code outside of the `server/` directory here? They are irrelevant. Putting them here looks like we are excluding those paths from the `/metrics` results as well.

Comment (aarnphm) [reply bentoml/_internal/server/metrics/prometheus.py:120]: imo `get_metrics_report` should filter out those paths from the report's perspective, and I'm ok with leaving those here.

File: bentoml/_internal/server/metrics/prometheus.py [hunk 7]
```diff
@@ -1,20 +1,39 @@
...
 logger = logging.getLogger(__name__)
 
+if TYPE_CHECKING:
+    from types import ModuleType
+
+    from ...external_typing import prometheus as ext
+
+
+# filter out unnecessary metrics
+EXCLUDE_PATHS = [
```
Comment (parano) [inline bentoml/_internal/server/metrics/prometheus.py:20]: also add "/" here?

File: bentoml/_internal/server/metrics/prometheus.py [hunk 8]
```diff
@@ -29,13 +48,14 @@ def __init__(
...
         self.multiproc = multiproc
         self.namespace = namespace
-        self.multiproc_dir: t.Optional[str] = multiproc_dir
-        self._registry = None
+        self.multiproc_dir: str | None = multiproc_dir
+        self._registry: ext.CollectorRegistry | None = None
         self._imported = False
-        self._pid: t.Optional[int] = None
+        self._pid: int | None = None
+        self._exc_paths: list[str] = EXCLUDE_PATHS
```
Comment (parano) [inline bentoml/_internal/server/metrics/prometheus.py:55]: remove, just use the constant below

File: bentoml/_internal/server/metrics/prometheus.py [hunk 9]
```diff
@@ -1,20 +1,39 @@
...
+
+
+# filter out unnecessary metrics
+EXCLUDE_PATHS = [
+    "/metrics",
+    "/static_content",
+    "/static_content/*",
+    "/docs.json",
+    "/openapi.json",
+    "/favicon.ico",
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply bentoml/_internal/server/metrics/prometheus.py:26]: Screenshot 2022-06-07 at 11 33 42

File: bentoml/_internal/server/metrics/prometheus.py [hunk 10]
```diff
@@ -83,41 +103,59 @@ def mark_process_dead(self) -> None:
...
-            return self.prometheus_client.generate_latest(registry)
+            return t.cast(bytes, self.prometheus_client.generate_latest(registry))
         else:
             return self.prometheus_client.generate_latest()
 
     def get_metrics_report(
-        self,
-    ) -> t.List[t.Dict[str, t.Union[str, float]]]:
-        metrics_text = self.generate_latest().decode()
+        self, filter_: list[str] | None = None
```
Comment (parano) [inline bentoml/_internal/server/metrics/prometheus.py:121]: remove filter_, not necessary

File: bentoml/_internal/server/metrics/prometheus.py [hunk 11]
```diff
@@ -83,41 +103,59 @@ def mark_process_dead(self) -> None:
...
                 and metric.name.endswith("_request")
             ):
-                return [
+                metric_ = [
                     {**sample.labels, "value": sample.value}
                     for sample in metric.samples
                 ]
-
-        return []
+                report_ += list(filter(truncate, metric_))
```
Comment (parano) [inline bentoml/_internal/server/metrics/prometheus.py:157]: it should just return here to stop looping through other metrics - we should have only one metrics matching the requirements above

File: bentoml/_internal/server/metrics/prometheus.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (parano) [reply bentoml/_internal/server/metrics/prometheus.py:26]: that's 404, not suppose to visit that

**Pull request closed without merging**